flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/EliminateCrossJoinRule.java

```diff
@@ -0,0 +1,477 @@
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.mapping.Mappings;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * This rule tries to eliminate cross joins by reordering joins.
+ * The new order of joins are determined with the following steps:
+ *
+ * <p>1. The inputs related with an equi-join filter (= or IS NOT DISTINCT FROM) will be joined first.
+ *       Inputs with smaller indices has higher priority.
+ *
+ * <p>2. The inputs related with other join filters will then be joined.
+ *
+ * <p>3. If not all inner join inputs are joined, they will be joined in input order.
+ *
+ * <p>4. Outer joins are added.
+ */
+public class EliminateCrossJoinRule extends RelOptRule {
+
+	public static final EliminateCrossJoinRule INSTANCE = new EliminateCrossJoinRule();
+
+	private EliminateCrossJoinRule() {
+		super(operand(MultiJoin.class, any()), "EliminateCrossJoinRule");
+	}
+
+	@Override
+	public void onMatch(RelOptRuleCall call) {
+		MultiJoin join = call.rel(0);
+		RelBuilder relBuilder = call.builder();
+
+		if (join.isFullOuterJoin()) {
+			// full outer join, do not reorder joins
+			Preconditions.checkArgument(
+				join.getInputs().size() == 2,
+				"Full outer join must have exactly 2 inputs. This is a bug.");
+
+			relBuilder
+				.push(join.getInput(0))
+				.push(join.getInput(1))
+				.join(JoinRelType.FULL, join.getJoinFilter());
+			if (join.getPostJoinFilter() != null) {
+				relBuilder.filter(join.getPostJoinFilter());
+			}
+		} else {
+			int outerJoinCount = 0;
+			for (int i = 0; i < join.getInputs().size(); i++) {
+				if (join.getJoinTypes().get(i) != JoinRelType.INNER) {
+					outerJoinCount++;
+				}
+			}
+			Preconditions.checkState(
+				outerJoinCount <= 1,
+				"EliminateCrossJoinRule assumes that there is at most 1 outer join " +
+					"in a layer of multi-join, but " + outerJoinCount + " outer joins were found.");
+			if (outerJoinCount == 1) {
+				int numInputs = join.getInputs().size();
+				Preconditions.checkState(
+					join.getJoinTypes().get(0) == JoinRelType.RIGHT ||
+						join.getJoinTypes().get(numInputs - 1) == JoinRelType.LEFT,
+					"EliminateCrossJoinRule assumes that " +
+						"the only left outer join input must locate at the end, or" +
+						"the only right outer join input must locate at the beginning");
+			}
+
+			LoptMultiJoin loptMultiJoin = new LoptMultiJoin(join);
+
+			// try to eliminate cross join
+			Vertex joinVertexTree = multiJoinToJoinVertexTree(loptMultiJoin);
+			Mappings.TargetMapping mapping = joinVertexTreeToJoinRelTree(joinVertexTree, loptMultiJoin, relBuilder);
+
+			// apply post-join filters
+			if (join.getPostJoinFilter() != null) {
+				RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+				relBuilder.filter(mapFilter(join.getPostJoinFilter(), mapping, rexBuilder));
```

Comment:
It's better to put `mapFilter(join.getPostJoinFilter(), mapping, rexBuilder)` part in a single line, which  is friendly to debug